harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,881 @@
+                    with self.prof.record_timing(
+                        f"callbacks.{callback.__class__.__name__}.on_training_start"
+                    ):
+                        callback.on_training_start()
+                self._run()
+
+        finally:
+            # Explicitly trigger the dataloader iterator shutdowns (which happens in __del__).
+            # See the rather long note in pytorch/torch/utils/data/dataloader.py.
+            if self.training_iterator is not None:
+                del self.training_iterator
+
+    def _run(self) -> None:
+        assert self.workloads is not None
+        for w, response_func in self.workloads:
+            try:
+                if w.kind == workload.Workload.Kind.RUN_STEP:
+                    action = "training"
+                    response = {
+                        "metrics": self._train_for_step(
+                            w.step_id,
+                            w.num_batches,
+                            w.total_batches_processed,
+                        ),
+                        "stop_requested": self.context.get_stop_requested(),
+                    }  # type: workload.Response
+
+                elif w.kind == workload.Workload.Kind.COMPUTE_VALIDATION_METRICS:
+                    action = "validation"
+                    response = {
+                        "metrics": self._compute_validation_metrics(),
+                        "stop_requested": self.context.get_stop_requested(),
+                    }
+
+                elif w.kind == workload.Workload.Kind.CHECKPOINT_MODEL:
+                    action = "checkpointing"
+                    # The checkpointing api would have been sufficient if the base_path for the
+                    # storage manager is guaranteed to be a shared file system.
+                    #
+                    # Since we can't guarantee that, we use the base storage_manager instead for
+                    # more flexibility.  Since checkpoints can be distributed across multiple
+                    # nodes, we will use the same uuid and separate path but each node
+                    # will upload its checkpoints to the storage manager individually.
+                    storage_manager = self.context._core.checkpointing._storage_manager
+                    if self.is_chief:
+                        metadata = {
+                            "latest_batch": self.latest_batch,
+                            "framework": f"torch-{torch.__version__}",
+                            "format": "pickle",
+                        }
+                        with storage_manager.store_path() as (
+                            storage_id,
+                            path,
+                        ):
+                            # Broadcast checkpoint path to all ranks.
+                            self.context.distributed._zmq_broadcast((storage_id, path))
+                            self._save(pathlib.Path(path))
```

Comment:
this looks to me like everybody's `self._save` gets the same path, and they're all going to write to the same files, and they'll all clobber each other.

Am I missing something?

Reply:
default deepspeed save will have ranks save different checkpoint files into the same directory to form a sharded checkpoint.

Reply:
This is unfortunate for us, but I suppose it is what it is until we upstream a better way.